Support 'string'-style queries on metadata fields when reasonable.

jtibshirani opened the pull request:
This change has two components:
- Support `prefix` and `regexp` queries on `_index` fields for consistency, since we already support `wildcard` queries.
- Update `ignored` and `routing` field types to inherit from `StringFieldType`, which allows them to support 'string'-style queries like `prefix`, `regex`, and `wildcard`. We temporarily removed support for `wildcard` as part of the refactor in #34062, and this PR restores that support for compatibility, and also makes sure the other query types are implemented for consistency.

Note that we may eventually deprecate support for 'string'-style queries on `ignored` and `routing` fields, pending a separate discussion.

jtibshirani commented on server/src/test/java/org/elasticsearch/index/mapper/IgnoredFieldTypeTests.java:
```diff
@@ -19,11 +19,25 @@
+
+        Query expected = new WildcardQuery(new Term("field", new BytesRef("foo*")));
+        assertEquals(expected, ft.wildcardQuery("foo*", null));
+    }
```
👍

jtibshirani commented on server/src/test/java/org/elasticsearch/index/mapper/RoutingFieldTypeTests.java:
```diff
@@ -18,12 +18,24 @@
+
+        Query expected = new WildcardQuery(new Term("field", new BytesRef("foo*")));
+        assertEquals(expected, ft.wildcardQuery("foo*", null));
+    }
```
👍

jtibshirani commented:
Sorry, should've explained that in the PR description. `_id` fields are actually encoded (see `IdFieldType#termsQuery`), so it didn't seem to make sense to support queries like `wildcard` + `prefix`.

jimczi commented:
Sorry I meant `IndexFieldType` for the `_index` field. It seems that we don't support `prefix` query on this field even though we have support for patterns in `IndexFieldType#termQuery` so it might be nice to fix.

jpountz commented:
Even though we could add support for these queries to these fields, should we? The `_id` field is a good example: if we had supported range/prefix/wildcard queries then we couldn't have optimized storage by detecting the format of these ids. Are there use-cases for prefix/wildcard queries on those fields? For instance I can't think of use-cases for searching the `_routing` field at all except maybe debugging, for which `term` queries are sufficient?

jpountz commented:
It looks like I might have commented too quickly and this PR is only restoring support that already existed before #34062?

jtibshirani commented:
@jimczi got it, I will look into adding support for the appropriate queries to `_index`.

@jpountz you're right, and I maybe should've done this as part of that PR to avoid confusion. This PR restores `wildcard` support to those field types, as well as adding support for the other 'string' query types like `prefix` for consistency.

jpountz commented:
Thanks @jtibshirani, I don't have objections then. +1 to discuss deprecating wildcard/prefix/range on those special fields.

jtibshirani commented:
@jimczi I think this is ready for another look. I added support for `prefix` queries to `_index` fields, and added more unit tests for the other types. I am actually a little unclear on why we want to add support for `prefix` to `_index` fields (given it's already a bit inconsistent as we don't support `regexp` or `fuzzy` queries).

I updated the title + PR description to reflect our discussion (and was being a bit indecisive about the title :))

jtibshirani commented:
Sounds good -- I'll file an issue to start the discussion around deprecation.
